File: swift/Sources/CantonDappKit/DappClient.swift
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// Moves JSON-RPC frames between a dApp and a wallet.
///
/// The currency is ``JSONRPCRequest``/``JSONRPCResponse`` rather than a typed
/// request union, because that is literally what crosses the wire and every
/// planned transport is a way of carrying those bytes somewhere: in-process,
/// deep link, LAN gRPC. Keeping the seam at the frame means a new transport
/// implements two members and inherits the whole protocol — and it is why the
/// JSON-RPC document stays the single schema.
public protocol DappTransport: Sendable {
    /// Sends a request and awaits its response.
    ///
    /// Throw ``DappError`` for protocol-level failures. A transport-level
    /// failure (socket closed, app not installed) may surface as any error;
    /// ``DappClient`` does not translate those, because a caller needs to tell
    /// "the wallet said no" apart from "the wallet was never reached".
    func send(_ request: JSONRPCRequest) async throws -> JSONRPCResponse

    /// Events pushed by the wallet.
    ///
    /// Defaults to an empty stream: a deep-link transport genuinely cannot
    /// deliver these, and forcing every implementation to write one would only
    /// obscure which ones can.
    var events: AsyncStream<DappEvent> { get }
}

extension DappTransport {
    public var events: AsyncStream<DappEvent> {
        AsyncStream { $0.finish() }
    }
}

/// The dApp side of CIP-0103: typed calls onto a ``DappTransport``.
///
/// ```swift
/// let client = DappClient(transport: transport)
/// if try await client.connect().isConnected {
///     let account = try await client.getPrimaryAccount()
///     let tx = try await client.prepareExecuteAndWait(request)
/// }
/// ```
///
/// Every method throws ``DappError`` when the wallet returns a JSON-RPC error,
/// so `4001` (the user declined) arrives as an error with `isUserRejection`
/// set rather than as a nil that is easy to mistake for "nothing happened".
///
/// Holds no session state. Whether a connection exists is the wallet's answer
/// to give, and caching it here would only let the two disagree.
public actor DappClient {
    private let transport: DappTransport
    private var nextId: Int64 = 1

    public init(transport: DappTransport) {
        self.transport = transport
    }

    /// Events pushed by the wallet, for transports that support them.
    public nonisolated var events: AsyncStream<DappEvent> { transport.events }

    /// Requests a connection. The wallet decides, typically by asking the user.
    public func connect() async throws -> ConnectResult {
        try DappJSON.decodeConnectResult(try await call(.connect))
    }

    /// Ends the session. Idempotent by convention.
    public func disconnect() async throws {
        _ = try await call(.disconnect)
    }

    /// Whether a connection exists, without prompting the user.
    public func isConnected() async throws -> ConnectResult {
        try DappJSON.decodeConnectResult(try await call(.isConnected))
    }

    /// Provider identity, connection state, and — if connected — network.
    public func status() async throws -> DappStatus {
        try DappJSON.decodeStatus(try await call(.status))
    }

    /// The network the wallet is on, as a CAIP-2 id plus optional endpoints.
    public func getActiveNetwork() async throws -> DappNetwork {
        try DappJSON.decodeNetwork(try await call(.getActiveNetwork))
    }

    /// Accounts the user granted this dApp — not every account the wallet holds.
    public func listAccounts() async throws -> [DappWallet] {
        try DappJSON.decodeAccounts(try await call(.listAccounts))
    }

    /// The account the wallet treats as primary, among those granted.
    public func getPrimaryAccount() async throws -> DappWallet {
        try DappJSON.decodeWallet(try await call(.getPrimaryAccount))
    }

    /// Asks the wallet to sign an arbitrary message. Subject to user approval.
    public func signMessage(_ message: String) async throws -> SignMessageResult {
        try DappJSON.decodeSignMessageResult(
            try await call(.signMessage, params: DappJSON.encode(SignMessageRequest(message: message)))
        )
    }

    /// Submits and returns as soon as the wallet accepts.
    ///
    /// Returns nothing — per OpenRPC this method's result is `Null`. The
    /// outcome arrives on ``events`` as `txChanged`, so this is only useful on
    /// a transport that has them; otherwise use ``prepareExecuteAndWait(_:)``.
    public func prepareExecute(_ request: PrepareSubmission) async throws {
        _ = try await call(.prepareExecute, params: DappJSON.encode(request))
    }

    /// Submits and waits until the transaction is executed.
    ///
    /// Throws if the user declines (`4001`) or the ledger rejects it
    /// (`-32003`); returns only on success, which is why the result is the
    /// executed event rather than a union.
    public func prepareExecuteAndWait(_ request: PrepareSubmission) async throws -> TxChangedEvent {
        try DappJSON.decodeExecutedResult(
            try await call(.prepareExecuteAndWait, params: DappJSON.encode(request))
        )
    }

    /// Calls the JSON Ledger API through the wallet, which supplies the
    /// credentials. Expect `4100` for resources outside the wallet's policy.
    public func ledgerApi(_ request: LedgerApiRequest) async throws -> JSONValue {
        try await call(.ledgerApi, params: DappJSON.encode(request))
    }

    /// Convenience for callers that want the raw frame — tests, diagnostics.
    public func request(method: String, params: JSONValue? = nil) async throws -> JSONRPCResponse {
        try await transport.send(JSONRPCRequest(method: method, params: params, id: takeId()))
    }

    private func call(_ method: DappMethod, params: JSONValue? = nil) async throws -> JSONValue {
        let response = try await transport.send(
            JSONRPCRequest(method: method.rawValue, params: params, id: takeId())
        )
        // The response id is deliberately not checked: some transports (a
        // deep-link callback, a multiplexed stream) legitimately reorder or
        // synthesise ids, and refusing the response would break them for no
        // safety gain — the transport is what guarantees pairing.
        return try response.resultOrThrow()
    }

    private func takeId() -> JSONValue {
        defer { nextId += 1 }
        return .int(nextId)
    }
}

File: kotlin/canton-dapp/src/main/kotlin/io/github/vsima/canton/dapp/DappClient.kt
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

package io.github.vsima.canton.dapp

import java.util.concurrent.atomic.AtomicLong
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive

/**
 * The dApp side of CIP-0103: typed calls onto a [DappTransport].
 *
 * ```kotlin
 * val client = DappClient(transport)
 * if (client.connect().isConnected) {
 *     val account = client.getPrimaryAccount()
 *     val tx = client.prepareExecuteAndWait(PrepareSubmission(commands = commands))
 *     println(tx.updateId)
 * }
 * ```
 *
 * Every method throws [DappException] when the wallet returns a JSON-RPC
 * error, so `4001` (the user declined) arrives as an exception with
 * [DappException.isUserRejection] set rather than as a null that is easy to
 * mistake for "nothing happened".
 *
 * This class holds no session state. Whether a connection exists is the
 * wallet's answer to give, and caching it here would only let the two
 * disagree — call [isConnected] or [status] to ask.
 */
public class DappClient(
    private val transport: DappTransport,
) {
    private val nextId = AtomicLong(1)

    /** Events pushed by the wallet, for transports that support them. */
    public val events: Flow<DappEvent> get() = transport.events

    /** Requests a connection. The wallet decides, typically by asking the user. */
    public suspend fun connect(): ConnectResult =
        DappJson.decodeConnectResult(call(DappMethod.CONNECT))

    /** Ends the session. Idempotent by convention; wallets should not error on a repeat. */
    public suspend fun disconnect() {
        call(DappMethod.DISCONNECT)
    }

    /** Whether a connection exists, without prompting the user. */
    public suspend fun isConnected(): ConnectResult =
        DappJson.decodeConnectResult(call(DappMethod.IS_CONNECTED))

    /** Provider identity, connection state, and — if connected — network and session. */
    public suspend fun status(): DappStatus =
        DappJson.decodeStatus(call(DappMethod.STATUS))

    /** The network the wallet is currently on, as a CAIP-2 id plus optional endpoints. */
    public suspend fun getActiveNetwork(): DappNetwork =
        DappJson.decodeNetwork(call(DappMethod.GET_ACTIVE_NETWORK))

    /** Accounts the user has granted this dApp. Not necessarily every account the wallet holds. */
    public suspend fun listAccounts(): List<DappWallet> =
        DappJson.decodeAccounts(call(DappMethod.LIST_ACCOUNTS))

    /** The account the wallet treats as primary, among those granted. */
    public suspend fun getPrimaryAccount(): DappWallet =
        DappJson.decodeWallet(call(DappMethod.GET_PRIMARY_ACCOUNT))

    /** Asks the wallet to sign an arbitrary message. Subject to user approval. */
    public suspend fun signMessage(message: String): SignMessageResult =
        DappJson.decodeSignMessageResult(
            call(DappMethod.SIGN_MESSAGE, DappJson.encode(SignMessageRequest(message))),
        )

    /**
     * Submits [request] and returns as soon as the wallet accepts it.
     *
     * Returns nothing — per OpenRPC this method's result is `Null`. The
     * outcome arrives on [events] as `txChanged`, so this overload is only
     * useful on a transport that has them; otherwise use
     * [prepareExecuteAndWait].
     */
    public suspend fun prepareExecute(request: PrepareSubmission) {
        call(DappMethod.PREPARE_EXECUTE, DappJson.encode(request))
    }

    /**
     * Submits [request] and suspends until the transaction is executed.
     *
     * Throws [DappException] if the user declines (`4001`) or the ledger
     * rejects it (`-32003`); returns only on success, which is why the
     * result type is the executed event rather than a union.
     */
    public suspend fun prepareExecuteAndWait(request: PrepareSubmission): TxChangedEvent.Executed =
        DappJson.decodeExecutedResult(
            call(DappMethod.PREPARE_EXECUTE_AND_WAIT, DappJson.encode(request)),
        )

    /**
     * Calls the JSON Ledger API through the wallet, which supplies the
     * credentials. The wallet applies its own policy — expect `4100` for
     * resources outside it rather than assuming full ledger access.
     */
    public suspend fun ledgerApi(request: LedgerApiRequest): JsonElement =
        call(DappMethod.LEDGER_API, DappJson.encode(request))

    private suspend fun call(method: DappMethod, params: JsonElement? = null): JsonElement {
        val id = JsonPrimitive(nextId.getAndIncrement())
        val response = transport.send(JsonRpcRequest(method.wire, params, id))
        // Not a hard failure: some transports (a deep-link callback, a
        // multiplexed stream) legitimately reorder or synthesise ids, and
        // refusing the response would break them for no safety gain — the
        // transport is what guarantees pairing.
        return response.resultOrThrow()
    }

    /** Convenience for callers that want the raw frame — mainly tests and diagnostics. */
    public suspend fun request(method: String, params: JsonElement? = null): JsonRpcResponse =
        transport.send(JsonRpcRequest(method, params, JsonPrimitive(nextId.getAndIncrement())))
}
